File: python/src/asqav/canonicalize.py
"""Fingerprint helpers producing the bytes the Asqav backend signs.

In hash-only mode the SDK hashes these same bytes locally. See
``docs/fingerprint-spec.md`` for canonical-form rules.
"""

from __future__ import annotations

import hashlib
import hmac
import json
from typing import Any

__all__ = [
    "canonicalize",
    "canonicalize_action",
    "canonicalize_tool_args",
    "hash_action",
]


def canonicalize_action(
    action_type: str,
    context: "dict[str, Any] | None" = None,
) -> bytes:
    """Canonical bytes of ``{action_type, context}`` for hashing or signing."""
    return canonicalize({"action_type": action_type, "context": context or {}})


def canonicalize(obj: Any) -> bytes:
    """Return JCS-subset JSON bytes for ``obj``; matches the conformance vectors byte-for-byte."""
    return json.dumps(
        obj,
        sort_keys=True,
        separators=(",", ":"),
        ensure_ascii=False,
        allow_nan=False,
    ).encode("utf-8")


def _find_float_pointer(value: Any, pointer: str) -> str | None:
    """Return JSON pointer to the first float in ``value`` (depth-first); ``bool`` is JCS-safe."""
    if isinstance(value, bool):
        return None
    if isinstance(value, float):
        return pointer
    if isinstance(value, dict):
        for key, sub in value.items():
            token = str(key).replace("~", "~0").replace("/", "~1")
            hit = _find_float_pointer(sub, f"{pointer}/{token}")
            if hit is not None:
                return hit
        return None
    if isinstance(value, list):
        for idx, sub in enumerate(value):
            hit = _find_float_pointer(sub, f"{pointer}/{idx}")
            if hit is not None:
                return hit
        return None
    return None


def canonicalize_tool_args(args: dict[str, Any] | list[Any]) -> bytes:
    """Return JCS-canonical bytes of ``tool_args``; raises ``ValueError`` (with JSON
    pointer) on any float leaf since floats aren't byte-stable across runtimes.
    """
    pointer = _find_float_pointer(args, "")
    if pointer is not None:
        raise ValueError(
            f"tool_args float at {pointer or '/'}: floats are not byte-stable "
            "across runtimes per RFC 8785 section 3.2.2; serialize numerics "
            "as strings or rationals before passing to canonicalize_tool_args"
        )
    return canonicalize(args)


def hash_action(
    action_type: str,
    context: dict[str, Any] | None = None,
    salt: bytes | None = None,
) -> str:
    """Return ``"sha256:<hex>"`` of canonicalized ``{action_type, context}``; with
    ``salt`` set, uses HMAC-SHA-256 keyed by that caller-held salt instead of plain
    SHA-256. The prefix reads ``sha256`` either way, so a verifier that recomputes
    per ``docs/fingerprint-spec.md`` cannot tell a salted digest from a plain one.
    """
    canonical = canonicalize_action(action_type, context)
    if salt is not None:
        digest = hmac.new(salt, canonical, hashlib.sha256).hexdigest()
    else:
        digest = hashlib.sha256(canonical).hexdigest()
    return f"sha256:{digest}"

File: typescript/src/canonicalize.ts
/**
 * Canonical JSON + content hashing for the @asqav/sdk TypeScript client.
 *
 * Mirrors the Python ``asqav.canonicalize`` module byte-for-byte. The
 * conformance vectors at ``conformance/vectors.json`` are validated against
 * both implementations, so a hash computed in Node and a hash computed in
 * Python over the same dict are guaranteed to agree.
 *
 * RFC 8785 (JCS) subset:
 *   - object keys sorted lexicographically
 *   - no whitespace anywhere
 *   - non-ASCII strings emitted as raw UTF-8
 *   - integers as integers (no trailing ``.0``)
 *   - finite floats only (NaN / Infinity rejected)
 */

import { createHash, createHmac } from "node:crypto";

export type JsonValue =
  | null
  | boolean
  | number
  | string
  | JsonValue[]
  | { [key: string]: JsonValue };

/**
 * Canonicalize a JSON-serializable value into UTF-8 bytes.
 *
 * Throws on values that aren't representable in JSON (functions, undefined,
 * BigInt, Symbol) and on non-finite numbers (NaN / Infinity), matching
 * the behavior of Python's ``json.dumps(allow_nan=False)``.
 */
export function canonicalize(value: unknown): Uint8Array {
  return new TextEncoder().encode(canonicalString(value));
}

function canonicalString(value: unknown): string {
  if (value === null) return "null";
  if (value === true) return "true";
  if (value === false) return "false";
  if (typeof value === "number") {
    if (!Number.isFinite(value)) {
      throw new Error("NaN / Infinity are not allowed in canonical JSON");
    }
    // Shortest round-trip representation per the canonical JSON spec via ECMAScript.
    return numberToCanonical(value);
  }
  if (typeof value === "string") {
    return jsonString(value);
  }
  if (Array.isArray(value)) {
    return "[" + value.map(canonicalString).join(",") + "]";
  }
  if (typeof value === "object") {
    const obj = value as Record<string, unknown>;
    const keys = Object.keys(obj).sort();
    const parts: string[] = [];
    for (const k of keys) {
      parts.push(jsonString(k) + ":" + canonicalString(obj[k]));
    }
    return "{" + parts.join(",") + "}";
  }
  throw new TypeError(`Not JSON-serializable: ${typeof value}`);
}

function numberToCanonical(n: number): string {
  // For integers within safe range, plain toString matches Python's repr.
  if (Number.isInteger(n) && Number.isSafeInteger(n)) {
    return n.toString();
  }
  return n.toString();
}

/**
 * RFC 8259 / RFC 8785 string serialization.
 *
 * Escapes control chars below U+0020, plus the two characters that JSON
 * requires (`"` and `\`). Everything U+0020 and above (including all
 * non-ASCII letters) is emitted as raw UTF-8. This matches Python's
 * ``json.dumps(ensure_ascii=False)`` which is what the conformance
 * vectors are generated with.
 */
function jsonString(s: string): string {
  let out = '"';
  for (let i = 0; i < s.length; i++) {
    const code = s.charCodeAt(i);
    const ch = s[i];
    if (ch === '"') {
      out += '\\"';
    } else if (ch === "\\") {
      out += "\\\\";
    } else if (code === 0x08) {
      out += "\\b";
    } else if (code === 0x09) {
      out += "\\t";
    } else if (code === 0x0a) {
      out += "\\n";
    } else if (code === 0x0c) {
      out += "\\f";
    } else if (code === 0x0d) {
      out += "\\r";
    } else if (code < 0x20) {
      out += "\\u" + code.toString(16).padStart(4, "0");
    } else {
      out += ch;
    }
  }
  out += '"';
  return out;
}

/** Canonical bytes of `{action_type, context}` for hashing or signing. */
export function canonicalizeAction(
  actionType: string,
  context: Record<string, unknown> = {},
): Uint8Array {
  return canonicalize({ action_type: actionType, context });
}

/**
 * Self-describing hash for an action: `sha256:<hex>`.
 * With `salt` set, uses HMAC-SHA-256.
 */
export async function hashAction(
  actionType: string,
  context: Record<string, unknown> = {},
  salt?: Uint8Array,
): Promise<string> {
  const canonical = canonicalizeAction(actionType, context);
  const hex = salt
    ? createHmac("sha256", Buffer.from(salt)).update(canonical).digest("hex")
    : createHash("sha256").update(canonical).digest("hex");
  return `sha256:${hex}`;
}
